File: sdk/packages/shared/src/persistence/DatabaseManager.ts
/**
 * DatabaseManager - TypeScript/Node.js SQLite Database Manager
 *
 * Central database manager for White Room persistence layer.
 * Manages SQLite database connection, WAL mode, and provides access to raw database operations.
 *
 * @module persistence
 * @author White Room Development Team
 * @copyright 2024 White Room. All rights reserved.
 */

import Database from 'better-sqlite3';
import * as fs from 'fs';
import * as path from 'path';

/**
 * Database-specific errors
 */
export enum DatabaseErrorCode {
  SCHEMA_FILE_NOT_FOUND = 'SCHEMA_FILE_NOT_FOUND',
  MIGRATION_FAILED = 'MIGRATION_FAILED',
  INITIALIZATION_FAILED = 'INITIALIZATION_FAILED',
  DATABASE_CORRUPTED = 'DATABASE_CORRUPTED',
  CONSTRAINT_VIOLATION = 'CONSTRAINT_VIOLATION',
}

export class DatabaseError extends Error {
  constructor(
    public code: DatabaseErrorCode,
    message: string,
    public originalError?: Error
  ) {
    super(message);
    this.name = 'DatabaseError';
  }
}

/**
 * Central database manager for White Room persistence layer.
 *
 * **Thread Safety:** better-sqlite3 is synchronous and thread-safe by default.
 * **Architecture:** Uses better-sqlite3 for direct SQLite access with prepared statements.
 *
 * @example
 * ```typescript
 * const dbManager = DatabaseManager.getInstance();
 * await dbManager.initialize();
 * const songCount = dbManager.read(db => {
 *   const stmt = db.prepare('SELECT COUNT(*) as count FROM songs');
 *   return stmt.get().count;
 * });
 * ```
 */
export class DatabaseManager {
  private static instance: DatabaseManager | null = null;
  private db: Database.Database;
  private dbPath: string;
  private isInitialized = false;

  /**
   * Private constructor (use getInstance() instead)
   */
  private constructor() {
    // Get database path
    const appSupportPath = this.getApplicationSupportDirectory();
    const databaseDir = path.join(appSupportPath, 'White Room');

    // Create directory if needed
    if (!fs.existsSync(databaseDir)) {
      fs.mkdirSync(databaseDir, { recursive: true });
    }

    // Set database path
    this.dbPath = path.join(databaseDir, 'white_room.db');

    // Create database connection
    try {
      this.db = new Database(this.dbPath);
      console.log(`[DatabaseManager] Database initialized at: ${this.dbPath}`);
    } catch (error) {
      throw new DatabaseError(
        DatabaseErrorCode.INITIALIZATION_FAILED,
        'Failed to initialize database',
        error as Error
      );
    }
  }

  /**
   * Get singleton instance
   *
   * @returns DatabaseManager instance
   */
  public static getInstance(): DatabaseManager {
    if (!DatabaseManager.instance) {
      DatabaseManager.instance = new DatabaseManager();
    }
    return DatabaseManager.instance;
  }

  /**
   * Initialize the database with schema and migrations
   *
   * This method should be called once at app startup. It:
   * 1. Enables WAL mode for concurrent access
   * 2. Runs any pending migrations
   * 3. Creates initial schema if needed
   *
   * @throws DatabaseError if initialization fails
   */
  public async initialize(): Promise<void> {
    try {
      // Setup WAL mode
      this.setupWALMode();
      console.log('[DatabaseManager] WAL mode enabled');

      // Run migrations
      await this.runMigrations();

      this.isInitialized = true;
      console.log('[DatabaseManager] Database initialization complete');
    } catch (error) {
      console.error('[DatabaseManager] Initialization failed:', error);
      throw error;
    }
  }

  /**
   * Get read-only access to the database
   *
   * Use this for SELECT queries and read operations.
   *
   * @param block - Closure receiving Database instance
   * @returns Result of the block
   *
   * @example
   * ```typescript
   * const song = dbManager.read(db => {
   *   const stmt = db.prepare('SELECT * FROM songs WHERE id = ?');
   *   return stmt.get(songId);
   * });
   * ```
   */
  public read<T>(block: (db: Database.Database) => T): T {
    return block(this.db);
  }

  /**
   * Get write access to the database
   *
   * Use this for INSERT, UPDATE, DELETE operations.
   * Automatically wraps the operation in a transaction.
   *
   * @param block - Closure receiving Database instance
   * @returns Result of the block
   *
   * @example
   * ```typescript
   * dbManager.write(db => {
   *   const stmt = db.prepare('INSERT INTO songs (id, name) VALUES (?, ?)');
   *   stmt.run(songId, songName);
   * });
   * ```
   */
  public write<T>(block: (db: Database.Database) => T): T {
    return this.db.transaction(block)();
  }

  /**
   * Check if database is initialized
   */
  public get initialized(): boolean {
    return this.isInitialized;
  }

  /**
   * Get the database file path
   */
  public get databasePath(): string {
    return this.dbPath;
  }

  /**
   * Get the database directory path
   */
  public get databaseDirectoryPath(): string {
    return path.dirname(this.dbPath);
  }

  /**
   * Setup WAL (Write-Ahead Logging) mode for better concurrent access performance
   *
   * WAL mode allows:
   * - Readers don't block writers
   * - Writers don't block readers
   * - Better concurrency for Node.js applications
   */
  private setupWALMode(): void {
    // Enable WAL mode
    this.db.pragma('journal_mode = WAL');

    // Use NORMAL synchronization (safer than OFF, faster than FULL)
    this.db.pragma('synchronous = NORMAL');

    // Set cache size to 64MB (-64000 means 64000KB)
    this.db.pragma('cache_size = -64000');

    // Store temp tables in memory
    this.db.pragma('temp_store = MEMORY');

    // Enable memory-mapped I/O (30GB max)
    this.db.pragma('mmap_size = 30000000000');

    // Enforce foreign key constraints
    this.db.pragma('foreign_keys = ON');

    console.log('[DatabaseManager] WAL mode configured');
  }

  /**
   * Run database migrations
   *
   * Checks current schema version and runs any pending migrations.
   */
  private async runMigrations(): Promise<void> {
    // Get current schema version
    const currentVersion = this.getSchemaVersion();
    console.log(`[DatabaseManager] Current schema version: ${currentVersion}`);

    // TODO: Run migrations based on version
    // For now, just ensure schema is created
    if (currentVersion === 0) {
      this.createInitialSchema();
      this.insertSchemaVersion(1);
    }
  }

  /**
   * Create initial database schema
   *
   * Executes the schema.sql file to create all tables.
   */
  private createInitialSchema(): void {
    console.log('[DatabaseManager] Creating initial schema...');

    // Try multiple paths for schema.sql
    const schemaPath = this.findSchemaFile();

    if (!schemaPath) {
      throw new DatabaseError(
        DatabaseErrorCode.SCHEMA_FILE_NOT_FOUND,
        'Schema file (schema.sql) not found'
      );
    }

    // Read schema file
    const schemaSQL = fs.readFileSync(schemaPath, 'utf-8');

    // Execute schema
    this.db.exec(schemaSQL);
    console.log('[DatabaseManager] Initial schema created');
  }

  /**
   * Find schema.sql file in multiple locations
   *
   * @returns Path to schema.sql or null if not found
   */
  private findSchemaFile(): string | null {
    const possiblePaths = [
      // Try relative to this file
      path.join(__dirname, '../../../../juce_backend/shared/persistence/schema.sql'),
      // Try from package root
      path.join(__dirname, '../../../juce_backend/shared/persistence/schema.sql'),
      // Try from project root
      path.join(process.cwd(), 'juce_backend/shared/persistence/schema.sql'),
      // Try from environment variable
      process.env.PROJECT_ROOT
        ? path.join(process.env.PROJECT_ROOT, 'juce_backend/shared/persistence/schema.sql')
        : null,
    ].filter(Boolean) as string[];

    for (const schemaPath of possiblePaths) {
      if (fs.existsSync(schemaPath)) {
        console.log(`[DatabaseManager] Found schema file: ${schemaPath}`);
        return schemaPath;
      }
    }

    return null;
  }

  /**
   * Insert schema version into metadata table
   *
   * @param version - Schema version number
   */
  private insertSchemaVersion(version: number): void {
    const stmt = this.db.prepare(
      'INSERT OR REPLACE INTO metadata (key, value, updated_at) VALUES (?, ?, datetime("now"))'
    );
    stmt.run('schema_version', version.toString());
    console.log(`[DatabaseManager] Schema version set to: ${version}`);
  }

  /**
   * Get current schema version from metadata table
   *
   * @returns Schema version number (0 if not set)
   */
  private getSchemaVersion(): number {
    const stmt = this.db.prepare('SELECT value FROM metadata WHERE key = ?');
    const result = stmt.get('schema_version') as { value: string } | undefined;
    return result ? parseInt(result.value, 10) : 0;
  }

  /**
   * Get the Application Support directory for the app
   *
   * @returns Path to ~/Library/Application Support/ (macOS) or equivalent
   */
  private getApplicationSupportDirectory(): string {
    const platform = process.platform;

    switch (platform) {
      case 'darwin': // macOS
        return path.join(process.env.HOME || '', 'Library', 'Application Support');

      case 'win32': // Windows
        return process.env.APPDATA || path.join(process.env.HOME || '', 'AppData', 'Roaming');

      case 'linux': // Linux
        return process.env.XDG_DATA_HOME || path.join(process.env.HOME || '', '.local', 'share');

      default:
        throw new Error(`Unsupported platform: ${platform}`);
    }
  }
}

/**
 * Database statistics utilities
 */
export class DatabaseStatistics {
  constructor(private dbManager: DatabaseManager) {}

  /**
   * Get database file size in bytes
   *
   * @returns File size in bytes
   */
  public databaseSize(): number {
    const stats = fs.statSync(this.dbManager.databasePath);
    return stats.size;
  }

  /**
   * Get database file size as human-readable string
   *
   * @returns Formatted file size (e.g., "12.5 MB")
   */
  public databaseSizeFormatted(): string {
    const bytes = this.databaseSize();
    const units = ['B', 'KB', 'MB', 'GB'];
    let size = bytes;
    let unitIndex = 0;

    while (size >= 1024 && unitIndex < units.length - 1) {
      size /= 1024;
      unitIndex++;
    }

    return `${size.toFixed(1)} ${units[unitIndex]}`;
  }

  /**
   * Get total number of songs in database
   *
   * @returns Song count
   */
  public songCount(): number {
    return this.dbManager.read((db) => {
      const stmt = db.prepare('SELECT COUNT(*) as count FROM songs');
      const result = stmt.get() as { count: number };
      return result.count;
    });
  }

  /**
   * Get total number of performances in database
   *
   * @returns Performance count
   */
  public performanceCount(): number {
    return this.dbManager.read((db) => {
      const stmt = db.prepare('SELECT COUNT(*) as count FROM performances');
      const result = stmt.get() as { count: number };
      return result.count;
    });
  }

  /**
   * Get current schema version
   *
   * @returns Schema version number
   */
  public schemaVersion(): number {
    return this.dbManager.read((db) => {
      const stmt = db.prepare('SELECT value FROM metadata WHERE key = ?');
      const result = stmt.get('schema_version') as { value: string } | undefined;
      return result ? parseInt(result.value, 10) : 0;
    });
  }

  /**
   * Perform database integrity check
   *
   * @returns true if database passes integrity check
   */
  public checkIntegrity(): boolean {
    return this.dbManager.read((db) => {
      const result = db.pragma('integrity_check', { simple: true });
      return result === 'ok';
    });
  }
}

/**
 * Debugging utilities (development only)
 */
export class DatabaseDebugger {
  constructor(private dbManager: DatabaseManager) {}

  /**
   * Log all table names and row counts
   */
  public debugLogTableCounts(): void {
    const tables = this.dbManager.read((db) => {
      const stmt = db.prepare("SELECT name FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%'");
      return stmt.all() as { name: string }[];
    });

    console.log('[DatabaseManager] === Database Statistics ===');
    for (const table of tables) {
      const count = this.dbManager.read((db) => {
        const stmt = db.prepare(`SELECT COUNT(*) as count FROM ${table.name}`);
        const result = stmt.get() as { count: number };
        return result.count;
      });
      console.log(`[DatabaseManager] ${table.name}: ${count} rows`);
    }
    console.log('[DatabaseManager] =============================');
  }

  /**
   * Export database table to JSON for debugging
   *
   * @param tableName - Name of table to export
   * @returns Array of objects representing rows
   */
  public debugExportTable(tableName: string): Record<string, any>[] {
    return this.dbManager.read((db) => {
      const stmt = db.prepare(`SELECT * FROM ${tableName}`);
      return stmt.all() as Record<string, any>[];
    });
  }
}

/**
 * Convenience function to get database statistics
 *
 * @param dbManager - DatabaseManager instance
 * @returns DatabaseStatistics instance
 */
export function getDatabaseStatistics(dbManager: DatabaseManager): DatabaseStatistics {
  return new DatabaseStatistics(dbManager);
}

/**
 * Convenience function to get database debugger (dev only)
 *
 * @param dbManager - DatabaseManager instance
 * @returns DatabaseDebugger instance
 */
export function getDatabaseDebugger(dbManager: DatabaseManager): DatabaseDebugger {
  return new DatabaseDebugger(dbManager);
}

File: swift_frontend/SwiftFrontendShared/Services/DatabaseManager.swift
//
//  DatabaseManager.swift
//  White Room
//
//  Created by White Room Development Team
//  Copyright © 2024 White Room. All rights reserved.
//

import Foundation
import GRDB

/// Central database manager for White Room persistence layer.
/// Manages SQLite database connection, WAL mode, and provides access to raw database operations.
///
/// **Thread Safety:** This class is thread-safe and can be accessed from multiple threads concurrently.
/// **Architecture:** Uses GRDB's DatabaseQueue for serialized access to SQLite database.
public final class DatabaseManager: Sendable {

    // MARK: - Singleton

    /// Shared singleton instance
    public static let shared = DatabaseManager()

    // MARK: - Properties

    /// GRDB DatabaseQueue for serialized database access
    private let dbQueue: DatabaseQueue

    /// File system URL to the database file
    private let databaseURL: URL

    /// Initialization lock for thread-safe lazy initialization
    private static let initLock = NSLock()

    /// Flag indicating if database has been initialized
    private var isInitialized = false

    // MARK: - Initialization

    /// Private initializer (use `DatabaseManager.shared` instead)
    private init() {
        // Get database path
        let appSupportURL = Self.applicationSupportDirectory
        let databaseDirectory = appSupportURL.appendingPathComponent("White Room", isDirectory: true)

        // Create directory if needed
        try? FileManager.default.createDirectory(at: databaseDirectory, withIntermediateDirectories: true)

        // Set database URL
        self.databaseURL = databaseDirectory.appendingPathComponent("white_room.db")

        // Create database queue
        do {
            self.dbQueue = try DatabaseQueue(path: self.databaseURL.path)
            print("[DatabaseManager] Database initialized at: \(self.databaseURL.path)")
        } catch {
            fatalError("[DatabaseManager] Failed to initialize database: \(error)")
        }
    }

    // MARK: - Public API

    /// Initialize the database with schema and migrations
    ///
    /// This method should be called once at app startup. It:
    /// 1. Enables WAL mode for concurrent access
    /// 2. Runs any pending migrations
    /// 3. Creates initial schema if needed
    ///
    /// - Throws: Database errors if initialization fails
    public func initialize() async throws {
        try await withCheckedThrowingContinuation { continuation in
            do {
                // Setup WAL mode
                try dbQueue.write { db in
                    try Self.setupWALMode(database: db)
                    print("[DatabaseManager] WAL mode enabled")
                }

                // Run migrations
                try await runMigrations()

                isInitialized = true
                print("[DatabaseManager] Database initialization complete")

                continuation.resume()
            } catch {
                print("[DatabaseManager] Initialization failed: \(error)")
                continuation.resume(throwing: error)
            }
        }
    }

    /// Get read-only access to the database
    ///
    /// Use this for SELECT queries and read operations.
    ///
    /// - Parameter block: Closure receiving Database instance
    /// - Throws: Database errors
    public func read<T>(_ block: (Database) throws -> T) throws -> T {
        return try dbQueue.read(block)
    }

    /// Get write access to the database
    ///
    /// Use this for INSERT, UPDATE, DELETE operations.
    ///
    /// - Parameter block: Closure receiving Database instance
    /// - Throws: Database errors
    public func write<T>(_ block: (Database) throws -> T) throws -> T {
        return try dbQueue.write(block)
    }

    /// Check if database is initialized
    public var initialized: Bool {
        return isInitialized
    }

    /// Get the database file path
    public var databasePath: String {
        return databaseURL.path
    }

    /// Get the database directory URL
    public var databaseDirectoryURL: URL {
        return databaseURL.deletingLastPathComponent()
    }

    // MARK: - Private Methods

    /// Setup WAL (Write-Ahead Logging) mode for better concurrent access performance
    ///
    /// WAL mode allows:
    /// - Readers don't block writers
    /// - Writers don't block readers
    /// - Better concurrency for iOS/macOS
    ///
    /// - Parameter database: GRDB Database instance
    /// - Throws: Database errors if PRAGMA execution fails
    private static func setupWALMode(database: Database) throws {
        // Enable WAL mode
        try database.execute(sql: "PRAGMA journal_mode = WAL")

        // Use NORMAL synchronization (safer than OFF, faster than FULL)
        try database.execute(sql: "PRAGMA synchronous = NORMAL")

        // Set cache size to 64MB (-64000 means 64000KB)
        try database.execute(sql: "PRAGMA cache_size = -64000")

        // Store temp tables in memory
        try database.execute(sql: "PRAGMA temp_store = MEMORY")

        // Enable memory-mapped I/O (30GB max)
        try database.execute(sql: "PRAGMA mmap_size = 30000000000")

        // Enforce foreign key constraints
        try database.execute(sql: "PRAGMA foreign_keys = ON")

        print("[DatabaseManager] WAL mode configured")
    }

    /// Run database migrations
    ///
    /// Checks current schema version and runs any pending migrations.
    ///
    /// - Throws: Database errors if migration fails
    private func runMigrations() async throws {
        try await withCheckedThrowingContinuation { continuation in
            do {
                try dbQueue.write { db in
                    // Get current schema version
                    let currentVersion = try Int.fetchOne(db, sql: "SELECT value FROM metadata WHERE key = 'schema_version'") ?? 0

                    print("[DatabaseManager] Current schema version: \(currentVersion)")

                    // TODO: Run migrations based on version
                    // For now, just ensure schema is created
                    if currentVersion == 0 {
                        try createInitialSchema(database: db)
                        try insertSchemaVersion(database: db, version: 1)
                    }
                }

                continuation.resume()
            } catch {
                print("[DatabaseManager] Migration failed: \(error)")
                continuation.resume(throwing: error)
            }
        }
    }

    /// Create initial database schema
    ///
    /// Executes the schema.sql file to create all tables.
    ///
    /// - Parameter database: GRDB Database instance
    /// - Throws: Database errors if schema creation fails
    private func createInitialSchema(database: Database) throws {
        print("[DatabaseManager] Creating initial schema...")

        // Read schema.sql file
        guard let schemaPath = Bundle.module.path(forResource: "schema", ofType: "sql") ??
                              Bundle.main.path(forResource: "schema", ofType: "sql") else {
            // Fallback: try to find schema.sql in juce_backend
            let projectRoot = ProcessInfo.processInfo.environment["PROJECT_ROOT"] ?? FileManager.default.currentDirectoryPath
            let fallbackPath = "\(projectRoot)/juce_backend/shared/persistence/schema.sql"

            guard FileManager.default.fileExists(atPath: fallbackPath) else {
                throw DatabaseError.schemaFileNotFound
            }

            // Read schema file
            let schemaSQL = try String(contentsOfFile: fallbackPath, encoding: .utf8)

            // Execute schema
            try database.execute(sql: schemaSQL)
            print("[DatabaseManager] Initial schema created from fallback path")

            return
        }

        // Read and execute schema file
        let schemaSQL = try String(contentsOfFile: schemaPath, encoding: .utf8)
        try database.execute(sql: schemaSQL)
        print("[DatabaseManager] Initial schema created")
    }

    /// Insert schema version into metadata table
    ///
    /// - Parameters:
    ///   - database: GRDB Database instance
    ///   - version: Schema version number
    /// - Throws: Database errors if insert fails
    private func insertSchemaVersion(database: Database, version: Int) throws {
        try database.execute(
            sql: "INSERT OR REPLACE INTO metadata (key, value, updated_at) VALUES (?, ?, datetime('now'))",
            arguments: ["schema_version", String(version)]
        )
        print("[DatabaseManager] Schema version set to: \(version)")
    }

    // MARK: - File System Utilities

    /// Get the Application Support directory for the app
    ///
    /// Returns: URL to ~/Library/Application Support/
    private static var applicationSupportDirectory: URL {
        let paths = FileManager.default.urls(
            for: .applicationSupportDirectory,
            in: .userDomainMask
        )

        guard let appSupport = paths.first else {
            fatalError("[DatabaseManager] Failed to get Application Support directory")
        }

        return appSupport
    }
}

// MARK: - Database Errors

/// Database-specific errors
public enum DatabaseError: LocalizedError {
    case schemaFileNotFound
    case migrationFailed(String)
    case initializationFailed(String)
    case databaseCorrupted
    case constraintViolation(String)

    public var errorDescription: String? {
        switch self {
        case .schemaFileNotFound:
            return "Schema file (schema.sql) not found"
        case .migrationFailed(let message):
            return "Migration failed: \(message)"
        case .initializationFailed(let message):
            return "Database initialization failed: \(message)"
        case .databaseCorrupted:
            return "Database file is corrupted"
        case .constraintViolation(let message):
            return "Database constraint violation: \(message)"
        }
    }
}

// MARK: - Database Statistics

extension DatabaseManager {

    /// Get database file size in bytes
    ///
    /// - Returns: File size in bytes
    /// - Throws: File system errors
    public func databaseSize() throws -> Int {
        let attributes = try FileManager.default.attributesOfItem(atPath: databaseURL.path)
        return attributes[.size] as? Int ?? 0
    }

    /// Get database file size as human-readable string
    ///
    /// - Returns: Formatted file size (e.g., "12.5 MB")
    public func databaseSizeFormatted() throws -> String {
        let bytes = try databaseSize()
        return ByteCountFormatter.string(fromByteCount: Int64(bytes), countStyle: .file)
    }

    /// Get total number of songs in database
    ///
    /// - Returns: Song count
    public func songCount() throws -> Int {
        return try dbQueue.read { db in
            try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM songs") ?? 0
        }
    }

    /// Get total number of performances in database
    ///
    /// - Returns: Performance count
    public func performanceCount() throws -> Int {
        return try dbQueue.read { db in
            try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM performances") ?? 0
        }
    }

    /// Get current schema version
    ///
    /// - Returns: Schema version number
    public func schemaVersion() throws -> Int {
        return try dbQueue.read { db in
            try Int.fetchOne(db, sql: "SELECT value FROM metadata WHERE key = 'schema_version'") ?? 0
        }
    }

    /// Perform database integrity check
    ///
    /// - Returns: true if database passes integrity check
    /// - Throws: Database errors
    public func checkIntegrity() throws -> Bool {
        return try dbQueue.read { db in
            let result = try String.fetchOne(db, sql: "PRAGMA integrity_check")
            return result == "ok"
        }
    }
}

// MARK: - Debugging Utilities

#if DEBUG

extension DatabaseManager {

    /// Log all table names and row counts (DEBUG only)
    public func debugLogTableCounts() throws {
        let tables = try dbQueue.read { db in
            try String.fetchAll(db, sql: "SELECT name FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%'")
        }

        print("[DatabaseManager] === Database Statistics ===")
        for table in tables {
            let count = try dbQueue.read { db in
                try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM \(table)") ?? 0
            }
            print("[DatabaseManager] \(table): \(count) rows")
        }
        print("[DatabaseManager] =============================")
    }

    /// Export database to JSON for debugging (DEBUG only)
    ///
    /// - Parameter tableName: Name of table to export
    /// - Returns: Array of JSON objects
    public func debugExportTable(_ tableName: String) throws -> [[String: Any]] {
        return try dbQueue.read { db in
            let rows = try Row.fetchAll(db, sql: "SELECT * FROM \(tableName)")
            return rows.map { row in
                var dict: [String: Any] = [:]
                for column in row.columnNames {
                    if let value = row[column] {
                        dict[column] = value
                    }
                }
                return dict
            }
        }
    }
}

#endif
